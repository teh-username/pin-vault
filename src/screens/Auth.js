import React from 'react';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';

import PasscodeAuth from '../components/PasscodeAuth';
import {
  getPasscodeRequirement,
  getCurrentPasscode,
} from '../redux/modules/settings';
import { hashString } from '../utils/crypto';

export class Auth extends React.Component {
  static navigationOptions = {
    title: 'Passcode Auth',
  };

  constructor(props) {
    super(props);
    this.state = {
      error: null,
    };
    this._isPasscodeValid = this._isPasscodeValid.bind(this);
    this._navToHome = this._navToHome.bind(this);
  }

  componentDidMount() {
    if (!this.props.requirePasscode) {
      this._navToHome();
    }
  }

  _navToHome() {
    this.props.navigation.navigate('Home');
    this.props.navigation.dispatch(
      NavigationActions.reset({
        index: 0,
        actions: [NavigationActions.navigate({ routeName: 'Home' })],
      })
    );
  }

  _isPasscodeValid(passcode) {
    const { currentPasscode } = this.props;
    if (hashString(passcode) !== currentPasscode) {
      this.setState({
        error: 'Passcode Invalid!',
      });
      return;
    }
    this._navToHome();
  }

  render() {
    const { error } = this.state;
    return (
      <PasscodeAuth
        handlePasscodeSubmit={this._isPasscodeValid}
        error={error}
      />
    );
  }
}

const mapStateToProps = state => ({
  requirePasscode: getPasscodeRequirement(state),
  currentPasscode: getCurrentPasscode(state),
});

export default connect(mapStateToProps)(Auth);
