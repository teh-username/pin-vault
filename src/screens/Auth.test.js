import React from 'react';

import { Auth } from './Auth';

describe('Auth screen test', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<Auth />, {
      disableLifecycleMethods: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should redirect to Home route if passcode protection is off', () => {
    const navMock = {
      navigate: jest.fn(),
      dispatch: jest.fn(),
    };
    const wrapper = shallow(
      <Auth requirePasscode={false} navigation={navMock} />
    );
    expect(navMock.navigate.mock.calls.length).toBe(1);
    expect(navMock.navigate.mock.calls[0][0]).toBe('Home');
    expect(navMock.dispatch.mock.calls[0][0]).toEqual({
      type: 'Navigation/RESET',
      index: 0,
      key: undefined,
      actions: [{ type: 'Navigation/NAVIGATE', routeName: 'Home' }],
    });
  });

  it('should not redirect to Home route if passcode protection is on', () => {
    const navMock = {
      navigate: jest.fn(),
      dispatch: jest.fn(),
    };
    const wrapper = shallow(
      <Auth requirePasscode={true} navigation={navMock} />
    );
    expect(navMock.navigate.mock.calls.length).toBe(0);
  });

  it('should throw an error if entered passcode is invalid and not redirect to Home', () => {
    const navMock = {
      navigate: jest.fn(),
      dispatch: jest.fn(),
    };
    const wrapper = shallow(
      <Auth
        requirePasscode={true}
        currentPasscode="5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8"
        navigation={navMock}
      />
    );

    wrapper.instance()._isPasscodeValid('notpassword');
    expect(wrapper.state('error')).toEqual('Passcode Invalid!');
    expect(navMock.navigate.mock.calls.length).toBe(0);
  });

  it('should redirect to Home if passcode entered is valid', () => {
    const navMock = {
      navigate: jest.fn(),
      dispatch: jest.fn(),
    };
    const wrapper = shallow(
      <Auth
        requirePasscode={true}
        currentPasscode="5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8"
        navigation={navMock}
      />
    );
    wrapper.instance()._isPasscodeValid('password');
    expect(wrapper.state('error')).toEqual(null);
    expect(navMock.navigate.mock.calls.length).toBe(1);
    expect(navMock.navigate.mock.calls[0][0]).toBe('Home');
    expect(navMock.dispatch.mock.calls[0][0]).toEqual({
      type: 'Navigation/RESET',
      index: 0,
      key: undefined,
      actions: [{ type: 'Navigation/NAVIGATE', routeName: 'Home' }],
    });
  });
});
