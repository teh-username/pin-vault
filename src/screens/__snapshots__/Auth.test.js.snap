// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth screen test should render as expected 1`] = `
<PasscodeAuth
  error={null}
  handlePasscodeSubmit={[Function]}
/>
`;
